### ISSUE TITLE
Enformer fine-tuning & benchmark data

### ISSUE THREAD

**Issue by shirondru** [Illumina/PromoterAI#8]
Hi there, thanks so much for your great work! I was wondering if you could share some more details about the Enformer fine-tuning procedure you report in figure S8, as the only information I could find is in the figure caption. Apologies if i missed it somewhere else! Any detail at all would be helpful, but some questions that immediately come to mind are:
1) what is the sequence length and learning rate used when fine-tuning enformer?
2) did you instantiate two copies of enformer for both the twin-net and concat-dense setups? If so, were there connections you established between the two copies whose weights were updated during fine-tuning?
3) Did you train on the exact same ~57,000 null variants + expression outliers?

Additionally, I noticed the benchmark data in the supplemental files and in this repository are missing the expression values. Would it be possible to share a version that includes that expression values for each genetic variant?

Thanks so much in advance for your help!

**Comment by fransilvion** [comment]
Hi, thank you so much for your interest in our work! I’m happy to provide the details here, please let me know if anything is unclear.

We fine-tuned a pre-trained Enformer with the official weights from DeepMind. The input sequence length was set to 49,152 bp, the learning rate to 5e-4, and the gradient norm clipping to 1. The batch size was 2 with 64 gradient accumulation steps. The model was trained for 150 epochs, with 1/5 of the data used in each epoch. Everything was run on a single A100 40GB GPU. And yes, we used the exact same null + expression outliers dataset.

For the actual model setup, I think it’s easier to show the code. For the concat-dense tuning configuration, we used the following class:

```python
class ConcatDense(nn.Module):# .
    def __init__(self, enformer,
                ):
        super().__init__()

        self.enformer = enformer
        self.fc1 = nn.Linear(3072*2, 1)

    def forward(self, input1, input2):
        output1 = self.enformer._trunk(input1)
        output2 = self.enformer._trunk(input2)

        outputs = torch.cat((output1, output2), -1)
        outputs = torch.mean(outputs, axis=1)

        output = self.fc1(outputs)

        return output
```
Here we collect the outputs from the Enformer trunk (the conv+transformer tower) for the reference sequence and the sequence with the alternative allele. We then concatenate them and process the result with a single dense layer. This dense layer is tuned together with the last layer of the trunk, while all other parameters are kept frozen:

```python
for param in enformer.parameters():
    param.requires_grad = False

for param in enformer._trunk[-1].parameters():
    param.requires_grad = True
```
Finally, the code for the twin-model set up:

```python
class TwinNet(nn.Module):# .
    def __init__(self, enformer,
                ):
        super().__init__()

        self.enformer = enformer

    def forward(self, input1, input2):
        output1 = self.enformer._trunk(input1)
        output2 = self.enformer._trunk(input2)

        output = output2 - output1
        output = torch.mean(output, axis=(1,2))
        output = self.tanh(output)

        return output
```

Here, just like above, we train only the last layer of the Enformer trunk. We also tried tuning the weights of the human output head, which gave similar results.

For the expression values, you can use the z_corrected column in Supplementary Table 2. This is the value we used for fine-tuning, it represents the normalized expression change due to the variant.

I hope this helps! Let me know if you have any other questions!

**Comment by shirondru** [comment]
Thank you so much! That code is very helpful. One more clarifying question about the enformer fine-tuning, did you use the same chromosome train/valid/test split that was used to fine-tune promoterAI, or did you stick to the original splits used to train Enformer?

For the expression values, I actually meant the benchmarks coming from datasets besides the rare promoter expression outliers. For example, the rows of the `MPRA_eQTL.tsv` ([here](https://github.com/Illumina/PromoterAI/blob/0c32ae77d6bbea565799d9531ae0c4f4c0c1d448/data/benchmark/MPRA_eQTL.tsv)) include variant positions but not the expression change attributed to each variant that you used in your benchmarks. I was hoping it might be possible to share a version of the files found under `data/benchmarks/` that include those expression values. For the `GTEx_outlier.tsv`, we can easily merge the values in like you say, but it's a bit more challenging to curate those values from the other sources.

Thanks again!

**Comment by fransilvion** [comment]
Got it, you can check the following files. Again, you can use Z_score column as the variant effect size. Let me know if you have questions!

[benchmark_GEL_RNA.tsv](https://github.com/user-attachments/files/21944087/benchmark_GEL_RNA.tsv)
[benchmark_GTex_eQTL.tsv](https://github.com/user-attachments/files/21944091/benchmark_GTex_eQTL.tsv)
[benchmark_GTEx_outlier.tsv](https://github.com/user-attachments/files/21944092/benchmark_GTEx_outlier.tsv)
[benchmark_MPRA_saturation.tsv](https://github.com/user-attachments/files/21944093/benchmark_MPRA_saturation.tsv)
[benchmark_UKBB_proteome.tsv](https://github.com/user-attachments/files/21944095/benchmark_UKBB_proteome.tsv)

**Comment by fransilvion** [comment]
And sorry, forgot to reply about the split, yes, we used the same train/test/validation split as for PromoterAI. We tried the split that was used for training Enformer and it was giving similar results. At the end, we report fine-tuned Enformer models with the best performance we could get.

**Comment by shirondru** [comment]
Thank you so much! You've been very helpful!
